### PR TITLE
[krel] Release Notes Editing Flow

### PR DESCRIPTION
**Pull request by puerco**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

This PR introduces the new release notes editing flow to krel.

It introduces a new flag `--fix` to the `krel release-notes` subcommand when creating the draft pull request that triggers the flow.

When `--fix` is set, krel goes into interactive mode, asking the user to review each release note that has to be reviewed. It will highlight any changes made to the note in the past. An example:
```
Release Note for PR 93107:
==========================
✨ Note contents are modified with a map
INFO Applying map to note from PR 93107           
    Author: bpineau
    SIGs: [cloud-provider]
    Kinds: [bug]
    Areas: [cloudprovider provider/azure]
 >> Feature: true
 >> ActionRequired: true
    ReleaseVersion: 1.19.0-rc.3
    Text:
    │ Azure: per VMSS VMSS VMs cache to prevent throttling on clusters having many
    │ attached VMSS
- Fix note for PR 93107? (y/N): 
```
If the user chooses to, krel will open their $EDITOR using kubectl's editor package, loading a release notes map,
preloaded with any changes, and with the original note commented for review. If the note contents are modified, krel will store the map yaml.

After each edit, progress is saved and the user may choose to exit the editing flow with ctrl+c keeping all progress made up to that point so that another member from the team can continue another time.

After editing is done, krel will regenerate the release notes draft, applying all the edits so far, commit it and push it to the release directory along with the maps and session files.

#### Which issue(s) this PR fixes:

Relates to #1373 

#### Special notes for your reviewer:
~For testing purposes, the pull request logic is temporarily commented out.~ I'm opening the PR for an initial review and ~will~ have re-enabled it along with the modifications to user input resulting from the changes coming in #1454 .

This PR depends on ~#1453~ and ~#1454~

#### Does this PR introduce a user-facing change?

```release-note
Release Notes Editing Flow. Adds the `--fix ` flag to the krel release-notes subcommand to enable the team to edit the release notes during the release cycle. It also adds logic to handle editing sessions and pushing the modified Draft back to k/sig-release
```
/assign @saschagrunert 
/cc @reylejano-rxm @JamesLaverack @wilsonehusin @AGMETEOR 

Signed-off-by: Adolfo García Veytia (Puerco) <adolfo.garcia@uservers.net>

### CODE REVIEW

File: cmd/krel/cmd/release_notes.go [hunk 1]
```diff
@@ -56,6 +61,13 @@ const (
 	websiteBranchPrefix = "release-notes-json-"
 	// draftBranchPrefix name of the draft branch in the user's fork
 	draftBranchPrefix = "release-notes-draft-"
+	// Editing instructions for notemaps
+	mapEditingInstructions = `# This is the current map for this Pull Request.
+# The original note content is commented out, if you need to
+# change a field, remove the comment and change the value.
+# To cancel, exit withouth changing anything or leave the file blank.
```
Comment (wilsonehusin) [inline cmd/krel/cmd/release_notes.go:68]: typo: withouth --> without 😬

File: cmd/krel/cmd/release_notes.go [hunk 2]
```diff
@@ -339,65 +370,141 @@ func createDraftPR(tag string) (err error) {
...
 
 	// generate the notes
-	targetdir := filepath.Join(sigReleaseRepo.Dir(), "releases", fmt.Sprintf("release-%d.%d", s.Major, s.Minor))
+	targetdir := filepath.Join(sigReleaseRepo.Dir(), "releases", fmt.Sprintf("release-%d.%d", tagVersion.Major, tagVersion.Minor))
 	logrus.Debugf("Release notes markdown will be written to %v", targetdir)
 	err = ioutil.WriteFile(filepath.Join(targetdir, draftFilename), []byte(result.markdown), 0644)
 	if err != nil {
 		return errors.Wrapf(err, "writing release notes draft")
 	}
+	logrus.Infof("Release Notes Draft written to %s", filepath.Join(targetdir, draftFilename))
```
Comment (wilsonehusin) [inline cmd/krel/cmd/release_notes.go:435]: ♥ for explicit logging!

File: cmd/krel/cmd/release_notes.go [hunk 3]
```diff
@@ -339,65 +370,141 @@ func createDraftPR(tag string) (err error) {
...
+	}
+
+	if !autoCreatePullRequest {
+		fmt.Println("Pull request creation aborted. Your local copy of k/sig-release")
+		fmt.Println("has note been deleted for you to review your changes. You may")
+		fmt.Println("push your changes to your fork and create the PR from there.")
+		fmt.Println("\nYour fork of kubernetes/sig-release was cloned here:")
+		fmt.Println(sigReleaseRepo.Dir())
+		logrus.Warn("Pull request aborted. Local changes were not pushed back.")
+		return nil
```
Comment (wilsonehusin) [inline cmd/krel/cmd/release_notes.go:452]: - do we favor this format over the multiline string?
- typo: "has **note** been deleted" --> "has **not** been deleted"
    - I think I would find it helpful with elaborate _what to do next_, something like:
      ```
      Pull request has NOT been created. The changes were made to your local copy of k/sig-release.
      To complete the process, you will need to:
        1. Review the changes in your local copy
        2. Push the changes to your fork (git push -u userfork <branch>)
        3. Submit a pull request to k/sig-release

      You can find your local copy here:
      /path/to/repository

      [WARN] Changes were made locally, user need to perform manual push and create pull request.
      ```
      what do you think? I'm raising this for discussion, not a grading rubric 😄

File: cmd/krel/cmd/release_notes.go [hunk 4]
```diff
@@ -339,65 +370,141 @@ func createDraftPR(tag string) (err error) {
...
+		prBody += "**Which issue(s) this PR fixes**:\n\n"
+		prBody += "**Special notes for your reviewer**:\n"
+		prBody += "This is an automated PR generated from `krel The Kubernetes Release Toolbox`\n\n"
+
+		// Create the pull request
+		logrus.Debugf(
+			"PR params: org: %s, repo: %s, headBranch: %s baseBranch: %s",
+			git.DefaultGithubOrg, git.DefaultGithubReleaseRepo, "master",
+			fmt.Sprintf("%s:%s", releaseNotesOpts.githubOrg, branchname),
+		)
```
Comment (wilsonehusin) [inline cmd/krel/cmd/release_notes.go:491]: is this supposed to be deleted?

Comment (wilsonehusin) [reply cmd/krel/cmd/release_notes.go:491]: womp nevermind I missed this comment:
> For testing purposes, the pull request logic is temporarily commented out. I'm opening the PR for an initial review and will re-enable it along with modifications to user input resulting from the changes coming in PR 1454 .

File: cmd/krel/cmd/release_notes.go (hunk 3 above)
Comment (puerco) [reply cmd/krel/cmd/release_notes.go:452]: Thanks @wilsonehusin this is precisely the suggestions I need: straight from the battle lines :) 
I like it better like you're suggesting, I'll add this in the next commit.

File: cmd/krel/cmd/release_notes.go [hunk 5]
```diff
@@ -787,11 +899,393 @@ func (o *releaseNotesOptions) Validate() error {
...
+		return nil, errors.Wrap(err, "reading working directory")
+	}
+	pullRequestList := make([]int, 0)
+
+	// Look in the work dir for all json files
+	for _, fileData := range files {
+		currentSession := &sessionData{}
+		if strings.HasSuffix(fileData.Name(), ".json") {
+			jsonData, err := ioutil.ReadFile(
+				fmt.Sprintf("%s%s%s", sessionPath,
```
Comment (JamesLaverack) [inline cmd/krel/cmd/release_notes.go:946]: `filepath.Join`?

File: cmd/krel/cmd/release_notes.go [hunk 6]
```diff
@@ -787,11 +899,393 @@ func (o *releaseNotesOptions) Validate() error {
...
+	files, err := ioutil.ReadDir(sessionPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading working directory")
+	}
+	pullRequestList := make([]int, 0)
+
+	// Look in the work dir for all json files
+	for _, fileData := range files {
+		currentSession := &sessionData{}
+		if strings.HasSuffix(fileData.Name(), ".json") {
```
Comment (JamesLaverack) [inline cmd/krel/cmd/release_notes.go:944]: We assume that any `.json` file in the working directory is a session file?

File: cmd/krel/cmd/release_notes.go [hunk 7]
```diff
@@ -787,11 +899,393 @@ func (o *releaseNotesOptions) Validate() error {
...
+			}
+			pullRequestList = append(pullRequestList, currentSession.PullRequests...)
+		}
+	}
+
+	// Copy the PRs to a map for easy lookup
+	pullRequestChecklist = map[int]bool{}
+	for _, pr := range pullRequestList {
+		pullRequestChecklist[pr] = true
+	}
```
Comment (JamesLaverack) [inline cmd/krel/cmd/release_notes.go:962]: What happens if a PR is mentioned multiple times? We'll overwrite the map with another one, and `ioutil.ReadDir` doesn't seem to have ordering guarantees. Could this lead to nondeterministic behaviour?

File: cmd/krel/cmd/release_notes.go [hunk 8]
```diff
@@ -787,11 +899,393 @@ func (o *releaseNotesOptions) Validate() error {
...
+	logrus.Infof("Read %d pull request reviews from previous sessions", len(pullRequestList))
+	return pullRequestChecklist, nil
+}
+
+// Do th fix process for the current tag
```
Comment (JamesLaverack) [inline cmd/krel/cmd/release_notes.go:967]:
```suggestion
// Do the fix process for the current tag
```

File: cmd/krel/cmd/release_notes.go [hunk 9]
```diff
@@ -787,11 +899,393 @@ func (o *releaseNotesOptions) Validate() error {
...
+		fmt.Println(spacer + strings.ReplaceAll(text, nl, nl+spacer))
+
+		choice, err := util.AskYorNorDefault(fmt.Sprintf("\n- Fix note for PR #%d?", note.PrNumber), false)
+		if err != nil {
+			// If the user cancelled with ctr+c exit and continue the PR flow
+			if err.(util.UserInputError).IsCtrlC() {
+				logrus.Info("Input cancelled, exiting edit flow")
+				return nil
+			}
+			return errors.Wrap(err, "while asking to edit release note")
```
Comment (JamesLaverack) [inline cmd/krel/cmd/release_notes.go:1091]: This only exits cleanly if you `Ctrl+C` while it's waiting for input? What if you happened to hit `Ctrl+C` while it wasn't? Would it be more robust to [use a goroutine to capture the signal interrupt instead](https://golangcode.com/handle-ctrl-c-exit-in-terminal/)?

File: cmd/krel/cmd/release_notes.go [hunk 10]
```diff
@@ -787,11 +899,393 @@ func (o *releaseNotesOptions) Validate() error {
...
+	}
+
+	// Remarshall the newyaml to save only the new values
+	newYAML, err := yaml.Marshal(testMap)
+	if err != nil {
+		return true, errors.Wrap(err, "while re-marshaling new map")
+	}
+
+	// Write the new map, removing the instructions
+	mapPath := fmt.Sprintf("%s%spr-%d-map.yaml", workdir, string(os.PathSeparator), pr)
```
Comment (JamesLaverack) [inline cmd/krel/cmd/release_notes.go:1283]: `filepath.Join`?

File: cmd/krel/cmd/release_notes.go (hunk 9 above)
Comment (puerco) [reply cmd/krel/cmd/release_notes.go:1091]: Yes, it only reacts when asking the user for input and it is built by handling the interrupt. Currently its part of #1454 
I'll move the signal handling code into the `util.Ask() ` function and use that function instead.

File: cmd/krel/cmd/release_notes.go (hunk 3 above)
Comment (puerco) [reply cmd/krel/cmd/release_notes.go:452]: Thanks for raising the point @JamesLaverack , I would have never known by myself that "aborted" should be avoided. Also glad to see wg-naming starting to have influence over what we're doing :)

Done, I've incorporated @wilsonehusin suggestion

File: cmd/krel/cmd/release_notes.go [hunk 11]
```diff
@@ -339,65 +370,141 @@ func createDraftPR(tag string) (err error) {
...
+		// Workdir is where we store the session state and all the release-notes maps
+		releaseDir := fmt.Sprintf(
+			"%s/releases/release-%d.%d", sigReleaseRepo.Dir(),
+			tagVersion.Major, tagVersion.Minor,
+		)
+
+		// Check if the release exists
+		if !util.Exists(releaseDir) {
+			return errors.New(fmt.Sprintf(
+				"could not find release directory for v%d.%d",
```
[the remark this answers is not in the document]
Comment (puerco) [reply cmd/krel/cmd/release_notes.go:386]: This one was, in fact, intentional. The directory names were determined years ago in the repository.  But everywhere we show the user a k8s version (or demand one like in cmd line args) we use the tag: v1.19.

I've changed the error to show the user the expected path to make the error more helpful.

File: cmd/krel/cmd/release_notes.go [hunk 12]
```diff
@@ -339,65 +370,141 @@ func createDraftPR(tag string) (err error) {
...
+		}
+
+		rnMapProvider, err := notes.NewProviderFromInitString(workdir)
+		if err != nil {
+			return errors.Wrap(err, "creating release notes draft")
+		}
+		for _, note := range releaseNotes.ByPR() {
+			maps, err := rnMapProvider.GetMapsForPR(note.PrNumber)
+			if err != nil {
+				return errors.Wrapf(err, "while getting maps for PR#%d", note.PrNumber)
```
[the remark this answers is not in the document]
Comment (puerco) [reply cmd/krel/cmd/release_notes.go:412]: This one should stay as is. If a map cannot be loaded, it probably has to do with a bad yaml file. It is worth giving the PR to the user to aid locating the file.

Also, krel processes PRs from k/k only. The general purpose release-notes tool has also support for maps, but this flow is only for kubernetes releases.

File: cmd/krel/cmd/release_notes.go [hunk 13]
```diff
@@ -339,65 +370,141 @@ func createDraftPR(tag string) (err error) {
...
+
+	// Generate the results struct
+	result, err := buildNotesResult(start, releaseNotes)
+	if err != nil {
+		return errors.Wrap(err, "building release notes results")
+	}
 
 	// generate the notes
-	targetdir := filepath.Join(sigReleaseRepo.Dir(), "releases", fmt.Sprintf("release-%d.%d", s.Major, s.Minor))
+	targetdir := filepath.Join(sigReleaseRepo.Dir(), "releases", fmt.Sprintf("release-%d.%d", tagVersion.Major, tagVersion.Minor))
```
[the remark this answers is not in the document]
Comment (puerco) [reply cmd/krel/cmd/release_notes.go:429]: yes, this was carried from earlier versions.  I've consolidated the two to releeaseDir

File: cmd/krel/cmd/release_notes.go [hunk 14]
```diff
@@ -339,65 +370,141 @@ func createDraftPR(tag string) (err error) {
...
 	logrus.Debugf("Release notes markdown will be written to %v", targetdir)
```
[the remark this answers is not in the document]
Comment (puerco) [reply cmd/krel/cmd/release_notes.go:430]: fixed, also came from previous versions

File: cmd/krel/cmd/release_notes.go [hunk 15]
```diff
@@ -339,65 +370,141 @@ func createDraftPR(tag string) (err error) {
...
 
-	// Create the PR
-	pr, err := gh.CreatePullRequest(
-		git.DefaultGithubOrg, git.DefaultGithubReleaseRepo, "master",
-		fmt.Sprintf("%s:%s", releaseNotesOpts.githubOrg, branchname),
-		fmt.Sprintf("Update release notes draft to version %s", tag), prBody,
-	)
+		// Create the PR
+		pr, err := gh.CreatePullRequest(
+			git.DefaultGithubOrg, git.DefaultGithubReleaseRepo, "master",
```
[the remark this answers is not in the document]
Comment (puerco) [reply cmd/krel/cmd/release_notes.go:495]: We have in fact a constant already named `git.Master` 
I've changed the code to use that one. Probably we'll have to rename that one too in the future

File: cmd/krel/cmd/release_notes.go (hunk 7 above)
Comment (puerco) [reply cmd/krel/cmd/release_notes.go:962]: We've maps override themselves in the order they come in. In the case of the krel flow it should not be a problem right now as no other map files should be in the directory, but it is definitely a good point to consider for the libraries in general. We should at least give them a predictable reading pattern.

File: cmd/krel/cmd/release_notes.go [hunk 16]
```diff
@@ -339,65 +370,141 @@ func createDraftPR(tag string) (err error) {
...
-	}()
+	// If we got the --fix flag, start the fix flow
+	if releaseNotesOpts.fixNotes {
+		// In interactive mode, we will ask the user before sending the PR
+		autoCreatePullRequest = false
+		// Workdir is where we store the session state and all the release-notes maps
+		releaseDir := fmt.Sprintf(
+			"%s/releases/release-%d.%d", sigReleaseRepo.Dir(),
+			tagVersion.Major, tagVersion.Minor,
+		)
```
[the remark this answers is not in the document]
Comment (puerco) [reply cmd/krel/cmd/release_notes.go:381]: Done, all have been replaced.

File: cmd/krel/cmd/release_notes.go (hunk 4 above)
Comment (puerco) [reply cmd/krel/cmd/release_notes.go:491]: PR logic has been re-added

File: cmd/krel/cmd/release_notes.go [hunk 17]
```diff
@@ -339,65 +370,141 @@ func createDraftPR(tag string) (err error) {
...
+		return nil
+	}
+
+	defer func() {
+		err = sigReleaseRepo.Cleanup()
+	}()
 
 	// add the updated file
-	if err := sigReleaseRepo.Add(filepath.Join("releases", fmt.Sprintf("release-%d.%d", s.Major, s.Minor), draftFilename)); err != nil {
+	if err := sigReleaseRepo.Add(filepath.Join("releases", fmt.Sprintf("release-%d.%d", tagVersion.Major, tagVersion.Minor), draftFilename)); err != nil {
```
[the remark this answers is not in the document]
Comment (puerco) [reply cmd/krel/cmd/release_notes.go:460]: Done, added a new var releasePath to hold the relative path while releaseDir has the full path to the repo

File: cmd/krel/cmd/release_notes.go [hunk 18]
```diff
@@ -339,65 +370,141 @@ func createDraftPR(tag string) (err error) {
...
-	// TODO: Maybe read and parse the PR template from sig-release?
-	prBody := "**What type of PR is this?**\n"
-	prBody += "/kind documentation\n\n"
-	prBody += "**What this PR does / why we need it**:\n"
-	prBody += fmt.Sprintf("This PR updates the Release Notes Draft to k/k %s\n\n", tag)
-	prBody += "**Which issue(s) this PR fixes**:\n\n"
-	prBody += "**Special notes for your reviewer**:\n"
-	prBody += "This is an automated PR generated from `krel The Kubernetes Release Toolbox`\n\n"
+		// push to the user's remote
+		logrus.Infof("Pushing modified release notes draft to %v/%v", releaseNotesOpts.githubOrg, releaseNotesOpts.draftRepo)
```
[the remark this answers is not in the document]
Comment (puerco) [reply cmd/krel/cmd/release_notes.go:471]: Don't know anymore as these came from older code shuffled around. Fixed.

File: cmd/krel/cmd/release_notes.go [hunk 19]
```diff
@@ -339,65 +370,141 @@ func createDraftPR(tag string) (err error) {
...
-	)
+		// Create a PR against k/sig-release using the github API
+		// TODO: Maybe read and parse the PR template from sig-release?
+		prBody := "**What type of PR is this?**\n"
+		prBody += "/kind documentation\n\n"
+		prBody += "**What this PR does / why we need it**:\n"
+		prBody += fmt.Sprintf("This PR updates the Release Notes Draft to k/k %s\n\n", tag)
+		prBody += "**Which issue(s) this PR fixes**:\n\n"
+		prBody += "**Special notes for your reviewer**:\n"
+		prBody += "This is an automated PR generated from `krel The Kubernetes Release Toolbox`\n\n"
```
[the remark this answers is not in the document]
Comment (puerco) [reply cmd/krel/cmd/release_notes.go:484]: I have a couple of plans like those, can we do those in follow up PRs?

File: cmd/krel/cmd/release_notes.go (hunk 6 above)
Comment (puerco) [reply cmd/krel/cmd/release_notes.go:944]: Yes all .json files in that directory are considered session files. It should not have other yaml or json files and they should not be allowed during the PR review.

File: cmd/krel/cmd/release_notes.go (hunk 7 above)
Comment (JamesLaverack) [reply cmd/krel/cmd/release_notes.go:975]: I guess we'd either want to order the files (so it's consistent at least), or do a map merge, or just error if there are multiples.

File: cmd/krel/cmd/release_notes.go (hunk 6 above)
Comment (JamesLaverack) [reply cmd/krel/cmd/release_notes.go:960]: Okay. Maybe a debug level log line to just print out the name of loaded files?

File: cmd/krel/cmd/release_notes.go [hunk 20]
```diff
@@ -339,32 +370,114 @@ func createDraftPR(tag string) (err error) {
...
+	// If we got the --fix flag, start the fix flow
+	if releaseNotesOpts.fixNotes {
+		// In interactive mode, we will ask the user before sending the PR
+		autoCreatePullRequest = false
+
+		workDir := filepath.Join(releaseDir, mapsDirectory)
+		if !util.Exists(workDir) {
+			logrus.Debugf("Working directory does not exist. Creating.")
+			if err := os.Mkdir(workDir, os.FileMode(0o755)); err != nil {
+				return errors.Wrap(err, "creating release notes workind directory")
```
Comment (JamesLaverack) [inline cmd/krel/cmd/release_notes.go:391]:
```suggestion
				return errors.Wrap(err, "creating release notes working directory")
```

File: pkg/notes/notes_map.go [hunk 21]
```diff
@@ -75,44 +75,45 @@ func ParseReleaseNotesMap(mapPath string) (*[]ReleaseNotesMap, error) {
...
-		Documentation *[]*Documentation `json:"documentation,omitempty"`
+		Documentation *[]*Documentation `json:"documentation,omitempty" yaml:"documentation,omitempty"`
 
 		// Author is the GitHub username of the commit author
-		Author *string `json:"author,omitempty"`
+		Author *string `json:"author,omitempty" yaml:"author,omitempty"`
 
 		// Areas is a list of the labels beginning with area/
-		Areas *[]string `json:"areas,omitempty"`
+		Areas *[]string `json:"areas,omitempty" yaml:"areas,omitempty"`
```
Comment (saschagrunert) [inline pkg/notes/notes_map.go:96]: Are those `yaml` tags really required? I thought they would be inferred from the `json` tags already.

File: cmd/krel/cmd/release_notes.go [hunk 22]
```diff
@@ -783,15 +911,395 @@ func (o *releaseNotesOptions) Validate() error {
...
+	userName, err := git.GetUserName()
+	if err != nil {
+		return errors.Wrap(err, "getting local user's name")
+	}
+
+	// Check the workDir before going further
+	dirData, err := os.Stat(workDir)
+	if os.IsNotExist(err) {
+		return errors.New("map directory does not exist")
+	}
```
Comment (saschagrunert) [inline cmd/krel/cmd/release_notes.go:997]: I think we have utility function in the `util` package for that.

File: pkg/notes/notes_map.go (hunk 21 above)
Comment (puerco) [reply pkg/notes/notes_map.go:96]: I thought so too. But after cracking my head for a while I managed to isolate the problem. It seems that the YAML output renders differently when values are missing when annotated with yaml: and json: 

I linked my test case in the comment: https://play.golang.org/p/C7QM1varozy
I do not know if its a bug or a feature (or maybe I'm doing something wrong that escapes me?)

Comment (saschagrunert) [reply pkg/notes/notes_map.go:96]: Hm that's indeed different :thinking:

**Comment by puerco** [comment]
Added a new feature to detect modifications when PRs are modified in Github. Now `--fix` will show the changed note to the user for them to review again.

File: cmd/krel/cmd/release_notes.go (hunk 6 above)
Comment (puerco) [reply cmd/krel/cmd/release_notes.go:980]: Added :+1:

File: cmd/krel/cmd/release_notes.go (hunk 20 above)
Comment (puerco) [reply cmd/krel/cmd/release_notes.go]: Fixed.

**Comment by puerco** [comment]
Final nits addressed. Now that #1454 has merged should be ready to go :rocket:

**Review by saschagrunert — APPROVED** [review]
/lgtm
/approve

**Pull request merged**